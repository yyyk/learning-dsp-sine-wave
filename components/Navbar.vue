<template>
  <nav class="nav">
    <div class="nav__button" @click="toggleNav">&#62; Table of Contents</div>
    <div :class="['nav__list-container', { active: active }]">
      <ul class="nav__list">
        <li v-for="m of menu" :key="m.path" class="list__item">
          <nuxt-link
            :to="m.path === '/index' ? '/' : m.path"
            exact-active-class="current"
            exact
          >
            {{ m.title }}
          </nuxt-link>
        </li>
      </ul>
    </div>
  </nav>
</template>

<script lang="ts">
import Vue from 'vue'
import { mapGetters } from 'vuex'

export default Vue.extend({
  data: () => ({
    active: false
  }),
  computed: {
    ...mapGetters(['menu'])
  },
  mounted() {
    this.$nuxt.$on('menu-close', () => {
      if (this.active) {
        this.active = false
      }
    })
  },
  beforeDestroy() {
    this.$nuxt.$off('menu-close')
  },
  methods: {
    toggleNav() {
      this.active = !this.active
    }
  }
})
</script>

<style lang="scss" scoped>
.nav {
  position: sticky;
  top: 0;
  width: 100%;
  background-color: $background-color;
  border-bottom: 1px solid #2c3e50;
  z-index: 10;
  padding: 16px;
  text-align: left;
  display: flex;
  flex-flow: column nowrap;
  align-items: center;
  justify-content: flex-start;
}

.nav__button {
  cursor: pointer;
}

.nav__list-container {
  max-height: 0;
  overflow: hidden;
  transition: max-height 0.4s ease-in-out;

  &.active {
    max-height: 100vh;
  }
}

.nav__list {
  list-style: none;
  margin: 24px 0 8px;

  .list__item {
    & + .list__item {
      margin-top: 16px;
    }

    a {
      text-decoration: none;

      &:hover {
        text-decoration: underline;
      }

      &.current {
        text-decoration: none;
        cursor: default;
        pointer-events: none;
        color: rgba(0, 0, 0, 0.5);
      }
    }
  }
}
</style>
