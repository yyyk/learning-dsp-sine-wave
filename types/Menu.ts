export interface MenuLink {
  title: string
  path: string
}
