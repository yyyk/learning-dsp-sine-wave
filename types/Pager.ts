export interface PageLink {
  path: string
  title?: string
  slug?: string
}
