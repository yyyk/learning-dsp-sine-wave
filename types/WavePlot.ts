export interface Sine {
  level: number
  ratio: number
  coarse: number
  fine: number
  frequencyOffset: number
}
