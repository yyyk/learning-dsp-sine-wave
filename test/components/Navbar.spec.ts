import Vuex from 'vuex'
import { mount, createLocalVue, RouterLinkStub } from '@vue/test-utils'
import Navbar from '@/components/Navbar.vue'
import { MenuLink } from '@/types/Menu'

const localVue = createLocalVue()
localVue.use(Vuex)

const store = new Vuex.Store({
  state: {
    menu: [
      {
        title: 'title1',
        path: '/index'
      },
      {
        title: 'title2',
        path: '/path2'
      },
      {
        title: 'title3',
        path: '/path3'
      }
    ] as MenuLink[]
  },
  getters: {
    menu: (state) => state.menu
  }
})

describe('Navbar.vue', () => {
  const mountFunction = (options = {}) => {
    return mount(Navbar, {
      store,
      localVue,
      stubs: { NuxtLink: RouterLinkStub },
      mocks: {
        $nuxt: {
          $on: () => ({}),
          $off: () => ({})
        }
      },
      ...options
    })
  }

  it('renders correctly', () => {
    const wrapper = mountFunction()
    expect(wrapper.html()).toMatchSnapshot()
  })

  it('exists', () => {
    const wrapper = mountFunction()
    expect(wrapper.exists()).toBe(true)
  })

  it('should have nav css class', () => {
    const wrapper = mountFunction()
    const parentDiv = wrapper.find('.nav')
    expect(parentDiv.exists()).toBe(true)
  })

  it('should have .nav__button', () => {
    const wrapper = mountFunction()

    const title = wrapper.find('.nav__button')
    expect(title.exists()).toBe(true)
  })

  describe('Nav menu', () => {
    it('should have menu closed by default', () => {
      const wrapper = mountFunction()
      expect(wrapper.vm.$data.active).toBe(false)
    })

    it('should open when .nav__button is clicked and toggleNav is called', async () => {
      const wrapper = mountFunction()
      wrapper.setData({
        active: false
      })
      const btn = wrapper.find('.nav__button')
      await btn.trigger('click')
      expect(wrapper.vm.$data.active).toBe(true)

      const el = wrapper.find('.nav__list-container')
      expect(el.classes().includes('active')).toBe(true)
    })

    it('should have .list__item if menu is not empty', () => {
      const wrapper = mountFunction()
      const links = wrapper.findAll('.nav__list .list__item')
      expect(links.length).toBeGreaterThan(0)
    })

    it("should have .link-text replace '/index' with '/'", () => {
      const wrapper = mountFunction()
      const link = wrapper
        .findAll('.nav__list .list__item')
        .at(0)
        .find('a')
      expect(link.props('to')).toBe('/')
    })
  })
})
