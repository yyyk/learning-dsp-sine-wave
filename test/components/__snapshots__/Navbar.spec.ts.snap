// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Navbar.vue renders correctly 1`] = `
"<nav class=\\"nav\\">
  <div class=\\"nav__button\\">&gt; Table of Contents</div>
  <div class=\\"nav__list-container\\">
    <ul class=\\"nav__list\\">
      <li class=\\"list__item\\"><a>
          title1
        </a></li>
      <li class=\\"list__item\\"><a>
          title2
        </a></li>
      <li class=\\"list__item\\"><a>
          title3
        </a></li>
    </ul>
  </div>
</nav>"
`;
